lib/Dialect/FIRRTL/Import/FIRParser.cpp

```diff
@@ -936,21 +941,96 @@ ParseResult FIRParser::importAnnotationFile(SMLoc loc) {
   return result;
 }
 
+Optional<std::pair<unsigned, unsigned>>
+FIRParser::getFieldIDFromTokens(SMLoc loc, ArrayAttr tokens, Type type) {
+  if (!type)
+    return None;
+  if (tokens.empty())
+    return std::pair<unsigned, unsigned>(0, 0);
+
+  auto currentType = type.cast<FIRRTLType>();
+  unsigned id = 0;
+  unsigned idRange = 0;
+
+  // Construct a string for error emission.
+  SmallString<16> tokensStr;
+  for (auto token : tokens) {
+    if (auto intAttr = token.dyn_cast<IntegerAttr>())
+      intAttr.getValue().toStringSigned(tokensStr);
+    else if (auto strAttr = token.dyn_cast<StringAttr>())
+      tokensStr.append(strAttr.getValue());
+    tokensStr.push_back('_');
+  }
+  tokensStr.pop_back();
```

Comment:
Can this not flatten the tokens and instead use `[`, `]`, and `.`? The error message could then be `foo.bar[0]` instead of `foo_bar_0`.